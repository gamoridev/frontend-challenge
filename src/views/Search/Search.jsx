import React from 'react';

import { SubHeader, SearchInput, Preview } from 'components'
import { SomosClient } from 'utils';
import { Heading, Alert } from 'plurall-ui';

import styles from './Search.module.css';
import noImage from '../../noImage.svg';

/* Handle user typing and key enter pressing */
const WAIT_INTERVAL = 500;

const client = new SomosClient()
class Search extends React.Component {
  state = {
    artists: [],
    notFound: false,
    loading: false,
    searchValue: '',
    typingTimeout: 0
  }

  handleChange = (value) => {
    const self = this;

    if (self.state.typingTimeout) {
      clearTimeout(self.state.typingTimeout);
    }

    if(value.length > 4){
      /* Timeout to handle user typing */
      self.setState({
        loading: true,
        searchValue: value,
        typingTimeout: setTimeout(function () {
          self.search(self.state.searchValue);
        }, WAIT_INTERVAL)
      });
    }
  }

  /* Use results from fetch */
  search = (value) => {
    client.getArtists(value)
    .then((res) => {
      if(res['artists'] && res['artists']['items'].length > 0){
        const artists = res['artists']['items'];
        this.setState({ loading: false, artists, notFound: false })
      } else {
        this.setState({ loading: false, artists: [], notFound: true })
      }
    })
    .catch((err) => {
      this.setState({ loading: false, notFound: false })
      return alert(err.message)
    });
  }

  render() {
    const { artists, notFound, searchValue, loading } = this.state;
    return (
      <React.Fragment>
        <SubHeader
          breadcrumb={[{ text: 'Pesquisa' }]}
          heading="Somos Front-end Challenge"
        />
        <div className={styles.wrapper}>
          <div className={styles.searchInput}>
            <SearchInput
              label='Encontre seus artistas preferidos'
              placeholder='Comece a digitar'
              notFoundMessage='No artists found'
              onChange={this.handleChange}
              value={searchValue}
            />
          </div>
          { artists.length > 0 ? 
            <React.Fragment>
              <Heading>{`Artistas encontrados: ${artists.length}`}</Heading>
              <section>
                {artists.map((e, k) => {
                  let background;
                  if (e['images'].length > 0){
                    background = e['images'][1]['url']
                  } else {
                    background = noImage;
                  }
                  return (
                    <Preview
                      key={k}
                      id={e['id']}
                      image={background}
                      name={e['name']}
                    />
                  )
                })}
              </section>
            </React.Fragment>
          : null }
          { loading ? <Heading>Carregando...</Heading> : null }
          { !loading && notFound ? <Heading>{`Nenhum artista encontrad@ para '${searchValue}'`}</Heading> : null }
        </div>
      </React.Fragment>
    )
  }
}

export default Search