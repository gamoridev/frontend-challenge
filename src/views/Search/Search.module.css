/* Workaround to use Search plurall component and remove results panel */
.searchInput div div div {
	display: none;
}

.wrapper{
	padding: 30px 20px;
}

.wrapper section{
	margin: 0 10px;
	display: flex;
	justify-content: space-around;
	flex-wrap: wrap;
}