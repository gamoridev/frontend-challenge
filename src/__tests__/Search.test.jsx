import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import Search from '../views/Search';


// TestRendering test
it('Search renders correctly', () => {
  const tree = renderer
    .create(<Router>
		  	<Search/>
			</Router>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});