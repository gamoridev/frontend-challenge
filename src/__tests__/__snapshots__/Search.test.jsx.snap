// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
Array [
  <div
    className="header"
  >
    <div
      className="wrapper"
    >
      <div
        className="innerDiv"
      >
        <div
          className="css-12qg8yb"
        >
          <div
            className="css-12qg8yb"
          >
            <div
              className="css-1xdrpc5"
            >
              <p
                className="css-1g0yeft"
              >
                Pesquisa
              </p>
            </div>
          </div>
        </div>
        <h1
          className="heading css-1n7fvwb"
        >
          Somos Front-end Challenge
        </h1>
      </div>
    </div>
  </div>,
  <div
    className="wrapper"
  >
    <div
      className="searchInput"
    >
      <div
        className="css-1iyoj2o"
      >
        <p
          className="css-stlxfv"
        >
          Encontre seus artistas preferidos
        </p>
        <div
          className="css-cssveg"
        >
          <input
            className="css-kxo4w0"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Comece a digitar"
            value=""
          />
          <div
            className="css-lwqyx4"
          >
            <div
              className="css-18oxwfi"
            >
              <span
                className="css-kr383g"
              >
                No artists found
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
